### ISSUE TITLE
op-chaining-support not loaded due to private field

### ISSUE THREAD

**Issue by muuvmuuv** [stephanrauh/ngx-extended-pdf-viewer#1812]
**Describe the bug**

Seems the latest feature test script does not work on older devices any more. Private fields are only available for chromium 74 and above, but we need to support older browsers/devices.

You can see the device infos below and can test with Android Studio Simulators or download old Chrome 66. Tested with our app and your website, https://pdfviewer.net/ (which als prints "Unexpected token ?").

I think because of not loading the op-chaining-support.js you aren't loading the legacy compiled es5> version, right?

**Suggestion**

Is the op-chaining-support really needed? I think a link to core-js to inform people to implement polyfills for their platform is enough, and a hint on that.

**Version info**
- Version of ngx-extended-pdf-viewer: 18
- Android 9 SDK 28
- WebView Chromium 66.0.3359.158

**Comment by muuvmuuv** [comment]
Or maybe something like this: https://replit.com/@muuvmuuv/FastHurtfulBruteforceprogramming#script.js

**Comment by muuvmuuv** [comment]
Oh, noticed that I can't use the es5 version anymore due to `??`... any ideas?
![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/11534760/d03e345c-df24-4430-8f7e-fa66df6deaa9)

**Comment by muuvmuuv** [comment]
Oh. PDF.js also uses a lot of `??=` and `||=` which are also not that backwards compatible :/
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment#browser_compatibility

**Comment by michaelmarcuccio** [comment]
I am getting a bunch of these errors as well. It started on 7/17/2023. I am using ngx-extended-pdf-viewer v16.2.3.

op-chaining-support.js:15:0
SyntaxError: Unexpected private name #privateMethod. Cannot parse class method with private name.

viewer-3.4.492-es5.min.js:44:232417
TypeError: undefined is not an object (evaluating 't.borderBoxSize[0]')

**Comment by stephanrauh** [comment]
@michaelmarcuccio It started two days ago? I didn't publish a version this week-end, let alone a 16.x. My best guess is that your error is a side-effect of the latest iOS security update 16.5.1 (c).

**Comment by stephanrauh** [comment]
@muuvmuuv @michaelmarcuccio I'm afraid I can't reproduce the bug. My iOS devices display the showcase without problems, my Safari too, and even on Browserstack I couldn't find a device reproducing the error. It seems the version you're referring to - Safari 14.1 and Chromium 66 - are simply too old for me. In other words, I need your help: How can I reproduce the bug?

**Comment by michaelmarcuccio** [comment]
The issue is due to incompatibility here: https://caniuse.com/mdn-javascript_classes_private_class_methods

The date this issue started for me was just the first time a user used an old browser with our app.

I cannot reproduce the bug either without the old browser, but there is a clear incompatibility listed in the link.

**Comment by stephanrauh** [comment]
Here's some background info: Marvin, the script you've mentioned works, but there are some caveats. First of all, I can't add it to my TypeScript code. Depending on your configuration, Angular generates two different binaries, so my TypeScript check would pass because it's cross-compiled to an older ECMAScript version. 

I can't merge it with any other JavaScript file. Some browsers allow us to catch the error, but other browsers don't. As far as I remember, Safari on iOS is one of the nasty ones. It simply stops with an error message. That's probably the first error message you, Michael, saw. I'd like to suppress it, but I don't know how to do it.

After a long research, one of my users came up with the idea of using a dedicated script to test compatibility. It's an ugly hack, it cost's a little performance, but it works flawlessly. Until two days ago, that is. 

However, I suspect something entirely different happened. I can confirm that the ES5 build of version 18 uses the ?? operator. It shouldn't do that - it's obviously an error. Maybe the pdf.js team has tweaked the transpilation options, and I didn't notice. And maybe they did it some time ago when I published version 16.4.3.

**Comment by stephanrauh** [comment]
@michaelmarcuccio The private method is meant to cause a bug. That's how I detect the browser needs the ES5 version of the library. Just ignore it and focus on the second error message. It shows that the browser detection worked. However, at second sight, it reads "undefined is not a function".

Digging deeper, I see that this is the relevant part of the sourcecode:
```js
  #resizeObserverCallback(entries) {
    for (const entry of entries) {
      if (entry.target === this.container) {
        this.#updateContainerHeightCss(
          Math.floor(entry.borderBoxSize[0].blockSize)
        );
        this.#containerTopLeft = null;
        break;
      }
    }
  }
```

Don't worry about the private method - your user's device runs the transpiled code. I'd like to show it, but to my dismay I noticed it's partially minified, so it's hard to read. Just in case you're curious, look here: https://unpkg.com/browse/ngx-extended-pdf-viewer@16.2.3/assets/viewer-3.4.492-es5.js

Be that as it may, I suspect you've reported a different bug than Marvin. Debugging taught me that `entry.target` is the `#viewerContainer` div. That's definitely not undefined - otherwise you'd reported a blank page or something like this.

Just an idea: is `ResizeObserverEntry.borderBoxSize` undefined on older Safaris?

**Comment by stephanrauh** [comment]
@michaelmarcuccio  Gotcha: CanIUse reports that `ResizeObserverEntry.borderBoxSize` is not supported on your user's device. I'm not sure I can do anything about it. I didn't find a polyfill. All I can do is adding a try-catch block. Adding insult to injury, I'd do it in version 17.4.x and 18.x only, so I'd force you to update.

<img width="1359" alt="grafik" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/a325c724-4ece-4268-8267-356b21554b9d">

**Comment by michaelmarcuccio** [comment]
Yep, that makes sense. Thanks for the insight on this. Do you know what this error/incompatibility does to the viewer's functionality? i.e. does it cause a crash or a bad behavior in the viewer? I am trying to reproduce it still.

**Comment by stephanrauh** [comment]
I'd called when the browser window is resized. In this case, there should be a minor crash. I can only guess: I expect the PDF viewer not to resize properly, but everything else should work properly. Unless the resize callback is called during the initialization phase, but I don't believe that.

**Comment by stephanrauh** [comment]
@muuvmuuv I've tried lowering the required versions, and it seems to work. Delivering the bugfix will take a few days. I've started to implement #1584, and I want to finish it before publishing a new version. Plus, I'd like to experiment with the settings. How bad is the price tag of transpiling? At the moment, the settings on my developer laptop look like so:

```js
const ENV_TARGETS = [
  "last 2 versions",
  "Chrome >= 66",  <<< down from 92
  "Firefox ESR",
  "Safari >= 14.1", <<< down from 15.4
  "Node >= 18",
  "> 1%",
  "not IE > 0",
  "not dead",
];
```

**Comment by muuvmuuv** [comment]
Here is our browserslistrc:

```json

  "browserslist": {
    "production": [
      "last 10 Chrome versions",
      "last 20 ChromeAndroid versions",
      "last 5 Safari versions",
      "last 3 iOS versions",
      "last 10 Firefox versions",
      "last 10 FirefoxAndroid versions",
      "last 5 Edge versions",
      "Firefox ESR",
      "not dead",
      "> 10% in DE"
    ],
    "mobile": [
      "iOS >= 13",
      "Android >= 9",
      "Chrome >= 66"
    ]
  },
```

in case you need it.

**Comment by muuvmuuv** [comment]
I think for the es5 build it is fine to be as backwards compatible as possible any users with a sense to tech will have a modern browser.

**Comment by stephanrauh** [comment]
You're right. Until recently, iOS needed the ES5 build, but that seems to be a thing of the past (at least for a while). :)

So I'm less reluctant to ship big JavaScript files for legacy browsers. On the other hand, that's not nice to people who're using a budget smartphone because that's all they can afford. Shipping files they can't run is even less fair, so I guess your browserslist.rc is a fair compromise.

**Comment by muuvmuuv** [comment]
Yeah thanks. We respect our customers with less bidget but dont want to stay too behind so we support OSsses three dears to the past. Next year it will be ios 14, and android 10. (i wish i could go to android 10 earlier... chrome 91...)

**Comment by stephanrauh** [comment]
Your bugfix has landed with version 17.4.6 and 18.0.0-alpha.5.

BTW: The impact on file size is remarkable: they almost doubled in size.

**Comment by muuvmuuv** [comment]
For web this might be an impact but not native Ionic like we use and anyway browser apps normaly dont need es5 anymore sooo imo this is a good middleway
